Improving Localization Coverage

indicozy opened the pull request:
## Proposed Changes
- Increased coverage of localization
  - [X] views (not mailers yet)
  - [ ] controllers (merged from @mabras's commits)
  - [ ] presenters (a bit)
  - [ ] ReScript (stuck at bug, no commits yet)
  Views are the majority of translations, I will go on improving localization coverage further. Please help me with translating ReScript, I am a bit stuck.
- 100% Translated `ru.yml` by analyzing `yml`s via [compare-yml](https://github.com/sony-mathew/compare-yml/)
- Fixed localization bugs such as `translation missing`

@pupilfirst/developers

## Merge Checklist
I don't know what to write there, I am new on contributing.

indicozy commented:
@mabras As I can see, in ReScript you used only `shared` hash from `I18n`. Do you know how to use other hashes? I cannot use anything except `shared`.

mabras commented:
> @mabras As I can see, in ReScript you used only `shared` hash from `I18n`. Do you know how to use other hashes? I cannot use anything except `shared`.

https://github.com/pupilfirst/pupilfirst/pull/928#issuecomment-1061595929

indicozy commented:
I don't know how to make `date-fns` locale-responsive. If anyone can help please reply
```rescript
                  <div className="text-xs">
                    <span> {t("issued_on")->str} </span>
                    <span className="ml-1">
                      {issuedCertificate
                      ->IssuedCertificate.createdAt
                      ->DateFns.formatPreset(~short=true, ~year=true, ())
                      ->str}
                    </span>
                  </div>
```

indicozy commented:
@mabras
> Is this working for you?
No, it wasn't working. I will make sure that those will work when I will finish it.
Thanks

indicozy commented:
Can I leave date range not like `date1 to date2` but `date1 - date2`

Photo1
![image](https://user-images.githubusercontent.com/51142471/157637520-cd002ad6-34c0-4bf4-b77e-b3b7828bdc39.png)

Photo2
![image](https://user-images.githubusercontent.com/51142471/157637356-a4c6bb54-aa23-49a3-be17-a84c715e921e.png)

mabras commented on db/migrate/20201109205227_add_locale_to_user.rb:
```diff
@@ -1,5 +1,5 @@
 class AddLocaleToUser < ActiveRecord::Migration[6.0]
   def change
-    add_column :users, :locale, :string, default: 'en'
+    add_column :users, :locale, :string, default: 'ru'
```
They will not accept this.

Maybe the following will help you:
```
class ApplicationController < ActionController::Base
  before_action :set_locale

  private

  def set_locale
    I18n.locale = :ru
  end
end
```

mabras replied:
@indicozy

indicozy replied:
I know, I messed up my github tree and merged my tinkers for my application. I will revert changes after I will finish translating

indicozy commented:
@mabras @harigopal
I've finished translation for 95-97% and I reverted unwanted commits. I need some help with testing and translating some details I've mentioned in the beginning:
1. For now, I changed many text that depended on pluralize and just pasted them to ymls in plural form. I'm not sure on how to make pluralize functionality without pluralize, please give some advises.
2. I couldn't find how notification widget works and how I can change header text in "Notice" and "Error" notifications.
3. I'm not sure how to make `date-fns` work with other localizations.

Also, it's time to test my code

mabras commented:
> I've finished translation for 95-97%

This great. Thanks!

> * For now, I changed many text that depended on pluralize and just pasted them to ymls in plural form. I'm not sure on how to make pluralize functionality without pluralize, please give some advises.

The rules of pluraization is difreent from one language to onother. First read this:
https://guides.rubyonrails.org/i18n.html#pluralization

Then check the following examples, in `ru`
https://github.com/svenfuchs/rails-i18n/blob/master/rails/locale/ru.yml#L65

example in `ar`
https://github.com/svenfuchs/rails-i18n/blob/master/rails/locale/ar.yml

The English ones are the simplest:
https://github.com/svenfuchs/rails-i18n/blob/master/rails/locale/en.yml

prashanth726 commented on config/locales/en.yml:
```diff
@@ -52,7 +100,7 @@ en:
         placeholder: Acme Inc., Acme University, etc.
       email:
-        label: Email
+        label: Emai l
```
```suggestion
        label: Email
```

prashanth726 commented on app/javascript/schools/communities/components/SchoolCommunities__Editor.res:
```diff
@@ -112,7 +115,7 @@ let handleQuery = (community, state, send, addCommunityCB, updateCommunitiesCB,
         }
 
-        Notification.success("Success", "Community updated successfully.")
+        Notification.success(ts("success"), ts("community_updated"))
```
```suggestion
        Notification.success(ts("notifications.success"), ts("notifications.community_updated"))
```

prashanth726 commented on app/javascript/schools/communities/components/SchoolCommunities__Editor.res:
```diff
@@ -1,5 +1,8 @@
 let str = React.string
 
+let ts = I18n.t(~scope="shared")
```
Please note : Scope referencing is wrong. 
```suggestion
let ts = I18n.t(~scope="shared.notifications")
```

indicozy commented:
Thanks, @prashanth726 ! I will later add DateTIme localization and work on notice, error notifications. It's better to leave `pluralize` as is for now, I can work on it in another PR.

indicozy commented:
Updated notifictions, @indicozytest is my second account, I pased wrong email in commits

indicozy commented:
@prashanth726 @mabras I've finished this commit, I will make other localization in other PRs in the future

harigopal commented on config/locales/en.yml:
```diff
@@ -1104,11 +1898,285 @@ en:
+        asked: asked %{days_ago} days ago.
+        control_emails: You can control these emails from %{link_to}.
+        link_to: your profile edit page
+  mailers:
+    applicant:
+      verify: Verify Your Email Address
+    coach:
+      added: You have been added as a coach in %{course_name}
+    course_author:
+      added: You have been added as an author in %{course_name}
+    course_export:
+      export: Export of %{course_export} course is ready for download
+    issued_certificate:
+      awarded: You have been awarded a certificate!
+    school_admin:
+      added: New School Admin Added
+      import: Import of Students Completed
+    startup:
+      new_feedback: New feedback from %{startup_feedback} on your submission
+    student:
+      added_as: You have been added as a student in %{school_name}
+    user:
+      new_reply: New reply for your post
+      daily_digest: Daily Digest
+      delete_account: Delete account from %{school_name}
+      deleted_success: Account deleted successfully from %{school_name}
+      deleted_in: Your account in %{school_name} will be deleted in 30 days
+    user_session:
+      log_in: Log in to
+      recovery: account recovery
```
All of these appear to be subjects of the corresponding mailer method. Could we stick to naming them by the caller mailer method? I.e., instead of `mailer.applicant.verify`, use `mailer.applicant.enrollment_verification_subject` so that it's clearer what the translation is supposed to be _of_.

prashanth726 replied:
Agreed @harigopal , I updated in this [commit ](https://github.com/pupilfirst/pupilfirst/pull/933/commits/9c6a71042082922b3ff0fa28b50031bb76cf26d4)

harigopal commented on app/views/issued_certificate_mailer/issued.html.erb:
```diff
@@ -7,21 +7,20 @@
 
   <p>
-    You've been awarded a certificate for completing the <strong><%= course.name %></strong> course.
+    <%= t("views.mailers.issued.awarded_certificate", course_name: course.name).html_safe %>
```
@pupilfirst/engineers I'm a bit concerned about the non-standard way we're handling interpolation here. I think to keep things safe, the _Rails-way_ is to have these keys that contain HTML be suffixed with `_html`, and  then to avoid using the `.html_safe` in the translation calls.

Here are the docs with examples: https://guides.rubyonrails.org/i18n.html#using-safe-html-translations

Using translated text that includes interpolation in this way should result in escaping only of the user-supplied values in the resulting text. That's what the docs say.

However, while this may work for Rails-rendered views, this approach won't work for client-side translations where such text would be injected into the DOM using `dangerouslySetInnerHTML`. I did a bit of looking up, and we may need to implement similar logic in the `I18n` module using DOMPurify. i.e., if the supplied key ends with `_html`, and if there are variables supplied for to the translation call, use DOMPurify to clean all variables before using it in the translation.

Here's a post in the `i18n-js` repo that includes an old implementation of this idea: https://github.com/fnando/i18n-js/issues/310#issuecomment-305909522

indicozy replied:
Thanks, I will update PR on the next week

mabras commented:
Related: https://github.com/pupilfirst/pupilfirst/issues/382

prashanth726 commented on app/javascript/schools/courses/components/curriculum_editor/CurriculumEditor__MarkdownBlockEditor.res:
```diff
@@ -12,5 +12,4 @@ let make = (~markdown, ~contentBlock, ~updateContentBlockCB) =>
     profile=Markdown.Permissive
     maxLength=10000
     onChange={onChange(contentBlock, updateContentBlockCB)}
-    dynamicHeight=true
```
@indicozy - This change is unrelated to this PR. Could you please stash or move them to a new PR?

prashanth726 commented on app/javascript/shared/components/MarkdownEditor.css:
```diff
@@ -6,15 +6,6 @@
 }
 
-.markdown-editor__fullscreen-preview-wrapper
-  > .markdown-editor__fullscreen-preview-editor {
+.markdown-editor__fullscreen-preview-wrapper>.markdown-editor__fullscreen-preview-editor {
   @apply max-w-3xl mx-auto;
 }
-
-.markdown-editor__textarea {
-  max-height: 12rem;
-}
-
-.markdown-editor__textarea--full-screen {
-  max-height: 100% !important;
-}
```
@indicozy - This file has changes unrelated to this PR. Could you please stash or move them to a new PR?

prashanth726 commented on app/javascript/shared/components/MarkdownEditor.res:
```diff
@@ -243,9 +234,9 @@ type phraseModifer =
 
 let insertAndWrapper = phraseModifer =>
   switch phraseModifer {
-  | Bold => (`**${t("bold_insert")}**`, "**")
-  | Italic => (`*${t("italic_insert")}*`, "*")
-  | Strikethrough => (`~~${t("strikethrough_insert")}~~`, "~~")
```
@indicozy - This file has changes unrelated to this PR. Could you please stash or move them to a new PR?

prashanth726 commented on app/javascript/shared/components/MarkdownEditor.res:
```diff
@@ -291,27 +282,18 @@ let controls = (disabled, value, state, send, onChange) => {
       <div />
     | Windowed(#Editor)
     | Fullscreen(#Editor | #Split) =>
-      <div role="toolbar" className="bg-white border border-gray-400 rounded-t border-b-0">
-        <button
-          disabled
-          ariaLabel={t("control_label_bold")}
-          title={t("control_label_bold")}
-          className=buttonClasses
-          onClick={_ => curriedModifyPhrase(Bold)}>
```
@indicozy - This file has lot of changes unrelated to this PR. Could you please stash or move them to a new PR?

prashanth726 commented on app/javascript/shared/components/Radio.css:
```diff
@@ -36,16 +36,10 @@
   @apply bg-primary-100 border-primary-500;
 }
 
-.radio-input {
-  top: 6px;
-  left: 6px;
-}
-
```
@indicozy - This file has changes unrelated to this PR. Could you please stash or move them to a new PR?

prashanth726 commented on app/javascript/shared/components/Radio.res:
```diff
@@ -4,8 +4,8 @@ let str = React.string
 
 @react.component
 let make = (~id, ~label, ~onChange, ~checked=false) =>
-  <div className="relative">
-    <input className="w-0 h-0 absolute radio-input focus:outline-none" id type_="radio" onChange checked />
```
@indicozy - This file has changes unrelated to this PR. Could you please stash or move them to a new PR?

prashanth726 commented on app/javascript/shared/components/MarkdownEditor.res:
```diff
@@ -275,12 +266,12 @@ let modifyPhrase = (oldValue, state, send, onChange, phraseModifer) => {
   }
 
 let controls = (disabled, value, state, send, onChange) => {
-  let buttonClasses = "px-2 py-1 hover:bg-gray-300 hover:text-primary-500 focus:outline-none focus:bg-gray-300 focus:text-primary-500 "
```
This change is unrelated to this PR.

prashanth726 reviewed (dismissed):
Hi @indicozy 

Thanks for all your work here. 👏

I have completed my review with this PR and made modifications wherever necessary, sharing my overall observations.  

1) I noticed some unrelated code changes related to the markdown editor in this PR, I commented on those files, can you remove them or move them to a new PR if necessary. 
2) The organization of keys in the `en.yml` needs to be improved  (for example, you have kept page-specific notifications in the shared section). I refactored the en.yml file, updated the missing keys and organized it according to the structure mentioned in our [developer documentation](https://developers.pupilfirst.com/docs/internationalization). You can update the `ru.yml` file as well based on this structure. 
3) You may notice lot of diff between the en.yml and ru.yml files because I changed the ordering of keys in the en.yml file using [pupilfirst translator vs code plugin](https://marketplace.visualstudio.com/items?itemName=bodhi.pupilfirst-translator), and I recommend using the translator plugin and order the keys when updating the ru.yml file as well.
4)  Also, we added a new  [Rspec matcher spec helper](https://github.com/pupilfirst/pupilfirst/pull/933/commits/7c21f2048790cb3800dcacbffc280a2f401702b2) for finding the missing keys in the UI in this PR.  This checks the page body for the indicators that the translation libraries attach such as `translation_missing` class that’s inserted by the regular Rails `t()` helper or the `[missing "en.whatever" translation]` text that i18n-js inserts. It runs after every js: true spec and displays the missing keys in the message. This makes fixing missing translation keys simple. 

Please let me know if you need any clarifications. Thanks again for all your work 🙂

harigopal commented on config/locales/en.yml:
```diff
@@ -868,36 +1595,108 @@ en:
       subscribed_on_another_device: Subscribe on this Device
       unsubscribe: Unsubscribe
+  controllers:
+    ApplicantsController:
+      link_expired: That one-time link has expired, or is invalid. If you have already completed enrollment, please sign in.
+      welcome: Welcome to %{school_name}!
+    CoursesController:
+      errors: "There were errors with your submission: %{form_errors}"
+      sent_mail: We've sent you a verification mail. It should reach you in less than a minute. Click the link in the email to sign up, and get started.
+    Schools__CoursesController:
+      teams_active: Teams marked active successfully!
+    UsersController:
+      link_expired: That link has expired or is invalid. Please try again.
+    Users__OmniauthCallbacksController:
+      add_facebook: Please <a href="https://www.facebook.com/settings?tab=applications" target="_blank" rel="noopener">remove \'Pupilfirst\' from your authorized apps list</a> and try signing in again.
+      add_github: Please <a href="https://github.com/settings/profile" target="_blank" rel="noopener">add a public email address to your Github profile</a> and try again.
+      add_other: Please sign in using another method.
+      denied: Authentication was denied. Please try again.
+      denied_by: Authentication was denied by %{provider}. Please try again.
+      email_unregistered: "Your email address: %{email} is unregistered."
+      not_receive_email: "We're sorry, but we did not receive your email address from %{provider_name}. "
```
@prashanth726 Could we not repeat _controllers_ in the key names here?

What I mean is, right now, it's `controllers.ApplicantsController.key`.

Instead, could we have `controllers.applicants.key`, or even better `controllers.applicants.ACTION.key`?

harigopal commented on config/locales/en.yml:
```diff
@@ -868,36 +1595,108 @@ en:
+      already_signed: You are already signed in.
+      link_expired: That one-time link has expired, or is invalid. Please try signing in again.
+      link_used: That one-time link has already been used, or is invalid. Please try resetting your password again.
   courses:
```
@prashanth726 This being controller keys should go under the new `controllers` namespace, right?

prashanth726 replied:
@harigopal - Yes, it is already in the controller namespace.  Did I miss something here?  🤔 
https://github.com/indicozy/op-onai-edu/blob/0922c691b3601d68b445ac3775ee30bb415a7d1f/config/locales/en.yml#L1627

harigopal replied:
@prashanth726 L1634 has the key `courses:` at an indentation of 2 spaces. L1200 has the key `controllers:` at an indentation of 2 spaces.

They're keys at the same level.

Also, there's a different key `courses:` inside `controllers:` at L1606.

harigopal commented on config/locales/en.yml:
```diff
@@ -868,36 +1595,108 @@ en:
       subscribe: Subscribe
       subscribed_on_another_device: Subscribe on this Device
       unsubscribe: Unsubscribe
+  controllers:
```
@prashanth726 Since this new key is being introduces, we'll also need to update the [internationalization guide](https://developers.pupilfirst.com/docs/internationalization/) in developer docs.

prashanth726 replied:
Done - https://github.com/pupilfirst/pupilfirst/pull/933/commits/0922c691b3601d68b445ac3775ee30bb415a7d1f

harigopal commented on docs/developers/docs/internationalization.md:
```diff
@@ -8,15 +8,16 @@ Pupilfirst LMS uses _Rails Internationalization_ (I18n) API for internationaliza
+| `mutations.MUTATION.*`            |            Strings used by mutations.             |
+| `queries.QUERY.*`                 |      Strings used by resolvers and mutators.      |
+| `shared.*`                        |                  Shared strings.                  |
+| `CONTROLLER.ACTION.*`             |        Strings used in traditional views.         |
```
@prashanth726 This last line is no longer required since the new L14 covers this.

mabras commented on config/locales/en.yml:
```diff
@@ -870,46 +1593,49 @@ en:
     messages:
-      content_type_invalid: has an invalid content type
-      limit_out_of_range: total number is out of range
+        content_type_invalid: has an invalid content type
+        limit_out_of_range: total number is out of range
```
Extra whitespace.

mabras commented on app/forms/users/sessions/sign_in_with_password_form.rb:
```diff
@@ -33,7 +33,10 @@ def shared_device?
-        errors.add(:base, I18n.t('sessions.create.invalid_credentials'))
+        errors.add(
+          :base,
+          'The supplied email address and password do not match. Please check your credentials and try again.'
```
Inline text.

mabras commented on config/locales/en.yml:
```diff
@@ -443,24 +599,26 @@ en:
-          failure: Some automated tests failed
-          error: Automated tests could not complete due to errors
       show_test_report_button: Show Test Report
       hide_test_report_button: Hide Test Report
```
Dup key.

mabras commented on config/locales/en.yml:
```diff
@@ -443,24 +599,26 @@ en:
           done_description: There are no other similar submissions.
       grade_card: Grade Card
       help_text: Notes can be used to keep track of a %{note_about}'s progress. These notes are shown only to coaches in a student's report. %{additional_help}
+      hide_test_report_button: Hide Test Report
```
Dup key.

mabras commented on app/javascript/users/UserEdit.res:
```diff
@@ -479,27 +465,24 @@ let make = (
                 id="language"
                 value={state.locale}
                 onChange={event => {
+                  Js.log(ReactEvent.Form.target(event)["value"])
```
Should be removed?

mabras commented on app/javascript/users/UserEdit.res:
```diff
@@ -479,27 +465,24 @@ let make = (
                   send(UpdateLocale(ReactEvent.Form.target(event)["value"]))
                 }}
-                className="select appearance-none block text-sm w-full bg-white shadow-sm border border-gray-400 rounded px-4 py-2 my-2 leading-relaxed focus:outline-none focus:border-transparent focus:ring-2 focus:ring-indigo-500">
+                className="appearance-none block text-sm w-full shadow-sm border border-gray-400 rounded px-4 py-2 my-2 leading-relaxed focus:outline-none focus:border-gray-500">
                 {availableLocales
                 ->Js.Array2.map(availableLocale =>
-                  <option key=availableLocale value=availableLocale ariaSelected={state.locale===availableLocale}>
+                  <option key=availableLocale value=availableLocale>
                     {Locale.humanize(availableLocale)->str}
```
Shouldn't be reverted?
